File: python/Q42_trapping_rain_water.py
#
# @lc app=leetcode id=42 lang=python3
#
# [42] Trapping Rain Water
#


# @lc code=start
class Solution:
    def trap(self, height: list[int]) -> int:
        if len(height) < 2:
            return 0

        left: int = 0
        right: int = len(height) - 1

        l_max: int = height[0]
        r_max: int = height[-1]
        volumn: int = 0

        while left <= right:
            if l_max < r_max:
                if (water := l_max - (new_max := height[left])) > 0:
                    volumn += water
                else:
                    l_max = new_max

                left += 1
            else:
                if (water := r_max - (new_max := height[right])) > 0:
                    volumn += water
                else:
                    r_max = new_max

                right -= 1

        return volumn


# @lc code=end

File: rust/src/leetcode/Q42_trapping_rain_water.rs
/*
 * @lc app=leetcode id=42 lang=rust
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut left_max: i32 = 0;
        let mut right_max: i32 = 0;

        let mut left: usize = 0;
        let mut right: usize = height.len() - 1;

        let mut result: i32 = 0;

        while left <= right {
            if left_max <= right_max {
                result += (left_max.min(right_max) - height[left]).max(0);
                left_max = left_max.max(height[left]);
                left += 1;
            } else {
                result += (left_max.min(right_max) - height[right]).max(0);
                right_max = right_max.max(height[right]);
                right -= 1;
            }
        }

        result
    }
}
// @lc code=end
